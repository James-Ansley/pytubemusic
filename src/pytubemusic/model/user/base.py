from pprint import PrettyPrinter

from pydantic import BaseModel, ConfigDict

__all__ = ("Model",)


class Model(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
        frozen=True,
    )


# noinspection PyUnresolvedReferences,PyProtectedMember
class _ModelPrinter(PrettyPrinter):
    def _pprint_model(self, obj, stream, indent, *args):
        stream.write(f"{type(obj).__name__}(")
        indent += len(type(obj).__name__) + 1
        non_nulls = {k: v for k, v in obj.__dict__.items() if v is not None}
        items = iter(non_nulls.items())
        k, v = next(items)
        stream.write(f"{k}=")
        self._format(v, stream, indent + len(k) + 1, *args)
        for k, v in items:
            stream.write(f"\n{' ' * indent}{k}=")
            self._format(v, stream, indent + len(k) + 1, *args)
        stream.write(f")")

    PrettyPrinter._dispatch[Model.__repr__] = _pprint_model
