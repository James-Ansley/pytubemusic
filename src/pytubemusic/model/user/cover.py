from pathlib import Path

from .base import Model

__all__ = (
    "Cover",
    "MaybeCover",
    "Url",
    "File",
)

type Cover = File | Url
type MaybeCover = Cover | None


class Url(Model):
    __match_args__ = ("href",)

    href: str


class File(Model):
    __match_args__ = ("path",)

    path: Path
