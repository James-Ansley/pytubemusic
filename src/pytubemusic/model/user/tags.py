from pytubemusic.model.types import MaybeStr, MaybeInt
from .base import Model

__all__ = (
    "Tags",
    "TrackTags",
    "AlbumTags",
    "MaybeTags",
    "MaybeTrackTags",
    "MaybeAlbumTags",
)

type MaybeTags = Tags | None
type MaybeTrackTags = TrackTags | None
type MaybeAlbumTags = AlbumTags | None


class Tags(Model):
    """FFMPEG Metadata tags"""

    __match_args__ = (
        "album",
        "title",
        "composer",
        "genre",
        "copyright",
        "encoded_by",
        "language",
        "artist",
        "album_artist",
        "performer",
        "disc",
        "publisher",
        "track",
        "encoder",
        "lyrics",
        "compilation",
        "date",
        "creation_time",
        "album_sort",
        "artist_sort",
        "title_sort",
    )

    title: MaybeStr = None
    album: MaybeStr = None
    composer: MaybeStr = None
    genre: MaybeStr = None
    copyright: MaybeStr = None
    encoded_by: MaybeStr = None
    language: MaybeStr = None
    artist: MaybeStr = None
    album_artist: MaybeStr = None
    performer: MaybeStr = None
    disc: MaybeStr = None
    publisher: MaybeStr = None
    track: MaybeStr | MaybeInt = None
    encoder: MaybeStr = None
    lyrics: MaybeStr = None
    compilation: MaybeStr = None
    date: MaybeStr = None
    creation_time: MaybeStr = None
    album_sort: MaybeStr = None
    artist_sort: MaybeStr = None
    title_sort: MaybeStr = None

    def __add__(self, other: "Tags") -> "Tags":
        return Tags(**(other.as_dict() | self.as_dict()))

    def as_dict(self) -> dict[str, int | str]:
        return {k: v for k, v in self.__dict__.items() if v is not None}


class TrackTags(Tags):
    """FFMPEG track metadata tags"""
    title: str
    album: MaybeStr = None


class AlbumTags(Tags):
    """FFMPEG album metadata tags"""
    album: str
    title: MaybeStr = None
