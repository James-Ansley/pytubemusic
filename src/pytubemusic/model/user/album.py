from pydantic import Field

from .base import Model
from .cover import MaybeCover
from .tags import AlbumTags
from .track import TrackType

__all__ = ("Album",)


class Album(Model):
    """A combination of individual tracks"""

    __match_args__ = ("metadata", "cover", "tracks")

    metadata: AlbumTags
    cover: MaybeCover = None
    tracks: tuple[TrackType, ...] = Field(min_length=1)
