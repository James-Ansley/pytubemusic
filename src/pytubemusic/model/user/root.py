from pydantic import RootModel

from .album import Album
from .track import TrackType

__all__ = ("MediaType", "Media")

type MediaType = TrackType | Album


class Media:
    def __new__(cls, **kwargs) -> MediaType:
        # noinspection PyTypeChecker
        return RootModel[MediaType](**kwargs).root
