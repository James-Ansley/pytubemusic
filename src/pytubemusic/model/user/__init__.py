"""
Data types for user-entered data
"""
from .album import *
from .cover import *
from .root import *
from .tags import *
from .track import *
