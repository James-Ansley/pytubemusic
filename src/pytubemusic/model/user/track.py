"""
Data types for collections of tracks
"""
from typing import Literal

from pydantic import Field, RootModel

from pytubemusic.model.types import MaybeTimedelta
from .base import Model
from .cover import MaybeCover
from .tags import TrackTags

__all__ = (
    "Track",
    "TrackType",
    "Single",
    "Split",
    "Playlist",
    "Merge",
    "TrackStub",
    "AudioStub",
    "TimeStub",
    "MergeStub",
    "Drop",
)

type TrackType = Single | Split | Playlist | Merge


class Track:
    def __new__(cls, **kwargs) -> TrackType:
        # noinspection PyTypeChecker
        return RootModel[TrackType](**kwargs).root


class Single(Model):
    """A single track"""

    __match_args__ = ("url", "metadata", "cover", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    metadata: TrackTags
    cover: MaybeCover = None
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TrackStub(Model):
    """A track that exists as a segment of a larger track"""

    __match_args__ = ("metadata", "cover", "start", "end")

    metadata: TrackTags
    cover: MaybeCover = None
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class Drop(Model):
    __match_args__ = ("drop",)

    drop: Literal[True] = True


class AudioStub(Model):
    """A track used only for audio with no associated metadata"""

    __match_args__ = ("url", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TimeStub(Model):
    """Timestamp information"""

    __match_args__ = ("url", "start", "end")

    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class MergeStub(Model):
    """Multiple time stamps clamped into one"""

    __match_args__ = ("metadata", "cover", "parts")

    metadata: TrackTags
    cover: MaybeCover = None
    parts: tuple[TimeStub | Drop, ...]


class Split(Model):
    """A single track that is split into smaller individual tracks"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/watch\?v=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub, ...] = Field(min_length=1)


class Playlist(Model):
    """A playlist"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/playlist\?list=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub | MergeStub | Drop, ...] = ()


class Merge(Model):
    """Multiple tacks combined into one"""

    __match_args__ = ("metadata", "cover", "tracks")

    metadata: TrackTags
    cover: MaybeCover = None
    parts: tuple[AudioStub, ...] = Field(min_length=1)
