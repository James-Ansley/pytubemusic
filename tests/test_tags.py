from pprint import pformat

from approvaltests import verify
from pydantic import ValidationError
from pytest import raises

from pytubemusic.model.user.tags import AlbumTags, Tags, TrackTags
from tests import test


@test()
def track_tags_can_be_created_with_minimal_data():
    data = {"title": "My Track Title"}
    tags = TrackTags(**data)
    assert isinstance(tags, TrackTags)
    assert tags.as_dict() == data


@test(depends_on=("track_tags_can_be_created_with_minimal_data",))
def track_tags_must_be_created_with_a_title():
    raises(ValidationError, lambda: TrackTags(**{}))
    raises(ValidationError, lambda: TrackTags(**{"genre": "Modern"}))
    raises(ValidationError, lambda: TrackTags(**{"album": "My Album Title"}))


@test()
def album_tags_can_be_created_with_minimal_data():
    data = {"album": "My Album Title"}
    tags = AlbumTags(**data)
    assert isinstance(tags, AlbumTags)
    assert tags.as_dict() == data


@test(depends_on=("album_tags_can_be_created_with_minimal_data",))
def album_tags_must_be_created_with_an_album():
    raises(ValidationError, lambda: AlbumTags(**{}))
    raises(ValidationError, lambda: AlbumTags(**{"genre": "Modern"}))
    raises(ValidationError, lambda: AlbumTags(**{"title": "My Track Title"}))


@test(
    depends_on=(
          "track_tags_can_be_created_with_minimal_data",
          "album_tags_can_be_created_with_minimal_data",
    )
)
def tags_can_be_merged():
    tags1 = TrackTags(**{"title": "My Track Title"})
    tags2 = AlbumTags(**{"album": "My Album Title"})
    tags = tags1 + tags2
    assert isinstance(tags, Tags)
    assert tags.as_dict() == {
        "title": "My Track Title",
        "album": "My Album Title",
    }


@test(depends_on=("tags_can_be_merged",))
def leftmost_tags_take_precedence_when_merging():
    tags1 = TrackTags(**{"title": "My Track Title", "genre": "Modern"})
    tags2 = AlbumTags(**{"album": "My Album Title", "genre": "Jazz"})
    tags3 = Tags(**{"title": "FooBar", "genre": "Classical"})
    tags = tags1 + tags2 + tags3
    assert isinstance(tags, Tags)
    assert tags.as_dict() == {
        "title": "My Track Title",
        "album": "My Album Title",
        "genre": "Modern",
    }


@test(
    depends_on=(
          "track_tags_can_be_created_with_minimal_data",
          "album_tags_can_be_created_with_minimal_data",
    )
)
def tags_are_pretty_printed():
    tags1 = TrackTags(**{"title": "My Track Title", "genre": "Modern"})
    tags2 = AlbumTags(**{"album": "My Album Title", "genre": "Jazz"})
    tags3 = Tags(**{"title": "FooBar", "genre": "Classical"})
    tag_fmt = "\n\n".join(pformat(tag) for tag in (tags1, tags2, tags3))
    verify(tag_fmt)
