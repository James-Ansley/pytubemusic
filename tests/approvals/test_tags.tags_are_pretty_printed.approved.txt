TrackTags(title='My Track Title'
          genre='Modern')

AlbumTags(album='My Album Title'
          genre='Jazz')

Tags(title='FooBar'
     genre='Classical')
